README
NAME
    Catalyst::Plugin::SmartURI - Configurable URIs for Catalyst

SYNOPSIS
    In your lib/MyApp.pm, load the plugin and your other plugins, for
    example:

        use Catalyst qw/
            -Debug
            ConfigLoader
            Static::Simple
            Session
            Session::Store::Memcached
            Session::State::Cookie
            Authentication
            Authorization::Roles
            +CatalystX::SimpleLogin
            SmartURI
        /;

    In your .conf:

        <Plugin::SmartURI>
            disposition host-header   # application-wide
            uri_class   URI::SmartURI # by default
        </Plugin::SmartURI>

    Per request:

        $c->uri_disposition('absolute');

    Methods on URIs:

        <a href="[% c.uri_for('/foo').relative %]" ...

DESCRIPTION
    Configure whether "$c->uri_for" and "$c->req->uri_with" return absolute,
    hostless or relative URIs, or URIs based on the 'Host' header. Also
    allows configuring which URI class to use. Works on application-wide or
    per-request basis.

    This is useful in situations where you're for example, redirecting to a
    lighttpd from a firewall rule, instead of a real proxy, and you want
    your links and redirects to still work correctly.

    To use your own URI class, just subclass URI::SmartURI and set
    "uri_class", or write a class that follows the same interface.

    This plugin installs a custom "$c->request_class", however it does so in
    a way that won't break if you've already set "$c->request_class"
    yourself, ie. by using Catalyst::Action::REST (thanks mst!).

    There is a minor performance penalty in perls older than 5.10, due to
    Class::C3, but only at initialization time.

METHODS
  $c->uri_for
  $c->req->uri_with
    Returns a "$c->uri_class" object (URI::SmartURI by default) in the
    configured "$c->uri_disposition".

  $c->req->uri
    Returns a "$c->uri_class" object. If the context hasn't been prepared
    yet, uses the configured value for "uri_class".

    "$c->req->uri->relative" will be relative to "$c->req->base".

  $c->req->referer
    Returns a "$c->uri_class" object for the referer (or configured
    "uri_class" if there's no context) with reference set to "$c->req->uri"
    if it comes from "$c->req->base".

    In other words, if referer is your app, you can do
    "$c->req->referer->relative" and it will do the right thing.

CONFIGURATION
    In myapp.conf:

        <Plugin::SmartURI>
            disposition absolute
            uri_class   URI::SmartURI
        </Plugin::SmartURI>

    disposition
        One of 'absolute', 'hostless', 'relative' or 'host-header'. Defaults
        to 'absolute'.

        The special disposition 'host-header' uses the value of your 'Host:'
        header.

    uri_class
        The class to use for URIs, defaults to URI::SmartURI.

PER REQUEST
        package MyAPP::Controller::RSSFeed;

        ...

        sub begin : Private {
            my ($self, $c) = @_;

            $c->uri_class('Your::URI::Class::For::Request');
            $c->uri_disposition('absolute');
        }

    $c->uri_disposition('absolute'|'hostless'|'relative'|'host-header')
        Set URI disposition to use for the duration of the request.

    $c->uri_class($class)
        Set the URI class to use for "$c->uri_for" and "$c->req->uri_with"
        for the duration of the request.

EXTENDING
    "$c->prepare_uri" actually creates the URI, which you can override.

SEE ALSO
    URI::SmartURI, Catalyst, URI

AUTHOR
    Rafael Kitover, "<rkitover at cpan.org>"

BUGS
    Please report any bugs or feature requests to
    "bug-catalyst-plugin-smarturi at rt.cpan.org", or through the web
    interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-Plugin-SmartURI
    >. I will be notified, and then you'll automatically be notified of
    progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Catalyst::Plugin::SmartURI

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-Plugin-SmartURI>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Catalyst-Plugin-SmartURI>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Catalyst-Plugin-SmartURI>

    *   Search CPAN

        <http://search.cpan.org/dist/Catalyst-Plugin-SmartURI>

ACKNOWLEDGEMENTS
    from #catalyst:

    vipul came up with the idea

    mst came up with the design and implementation details for the current
    version

    kd reviewed my code and offered suggestions

TODO
    I'd like to extend on Catalyst::Plugin::RequireSSL, and make a plugin
    that rewrites URIs for actions with an SSL attribute.

COPYRIGHT & LICENSE
    Copyright (c) 2008 Rafael Kitover

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Catalyst-Plugin-SmartURI

0.039 2014-01-11 15:55:01
    - remove Class::MOP::load_class RT#90916

0.038 2013-11-14 20:02:08
    - return undef for $c->req->referer correctly when there is no Referer
      header

0.037 2013-08-26 23:50:44
    - converted to git and moved to github
    - converted to dzil
    - add plugin loading example to SYNOPSIS (RT#88069)

0.036  2009-11-24 09:08:09
    - POD fixes from Frank Wiegand (RT#51885)

0.035  2009-09-24 04:35:28
    - fix typo in pod (RT#49933)

0.034  2009-05-13 23:46:18
    - fix compatibility with Catalyst::Action::REST v0.71
    - move inline tests into t/lib

0.033  2009-04-29 19:46:04
    - last dist was missing inc/

0.032  2009-04-27 15:42:07
    - converted to Moose
    - some minor POD cleanups

0.030  2008-11-19 17:54:37
    - fixed for Catalyst 5.8

0.029  2008-10-09 07:02:46
    - fixed the host-header stuff for when there's no port.

0.028  2008-08-22 03:53:49
    - added note to the POD about usage with C::C::REST.

0.027  2008-07-06 14:17:12
    - improved request class, better Host header parsing, fixed POD.

0.025  2008-06-28 08:33:57
    - fucked up the pod for SYNOPSIS...

0.024  2008-06-28 03:01:53
    - added 'host-header' disposition.
    - converted conf key to 'Plugin::SmartURI'

0.023  2008-06-28 01:14:49
    - weaken saved context.

0.022  2008-06-24 16:52:55
    - fix weird bug with initialization that sometimes happens.

0.021  2008-06-16 22:20:42
    - minor speedup, avoiding unnecessary string evals.

0.02  2008-06-03 21:58:17
    - moved Catalyst::SmartURI into a separate dist as URI::SmartURI

0.01_01  2008-04-01 08:33:13
    - first dev release.

0.01  2008-03-08 03:30:31
    - first version.


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.008.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Configurable URIs for Catalyst",
  "AUTHOR" => "Rafael Kitover <rkitover\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catalyst-Plugin-SmartURI",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::Plugin::SmartURI",
  "PREREQ_PM" => {
    "Catalyst::Exception" => 0,
    "Catalyst::Request" => 0,
    "Class::C3::Componentised" => 0,
    "Class::Load" => 0,
    "Moose" => 0,
    "Scalar::Util" => 0,
    "mro" => 0,
    "namespace::clean" => 0
  },
  "TEST_REQUIRES" => {
    "Catalyst" => 0,
    "Catalyst::Action::REST" => 0,
    "Catalyst::Controller" => 0,
    "Catalyst::Test" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "HTTP::Request" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0,
    "URI::SmartURI" => 0,
    "lib" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.039",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Catalyst" => 0,
  "Catalyst::Action::REST" => 0,
  "Catalyst::Controller" => 0,
  "Catalyst::Exception" => 0,
  "Catalyst::Request" => 0,
  "Catalyst::Test" => 0,
  "Class::C3::Componentised" => 0,
  "Class::Load" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "HTTP::Request" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moose" => 0,
  "Scalar::Util" => 0,
  "Test::More" => 0,
  "URI::SmartURI" => 0,
  "lib" => 0,
  "mro" => 0,
  "namespace::clean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Catalyst/Plugin/SmartURI.pm
package Catalyst::Plugin::SmartURI;
BEGIN {
  $Catalyst::Plugin::SmartURI::AUTHORITY = 'cpan:RKITOVER';
}
{
  $Catalyst::Plugin::SmartURI::VERSION = '0.039';
}

use Moose;
use mro 'c3';

use 5.008001;
use Class::C3::Componentised;
use Scalar::Util 'weaken';
use Catalyst::Exception ();
use Class::Load ();

use namespace::clean -except => 'meta';

has uri_disposition => (is => 'rw', isa => 'Str');
has uri_class       => (is => 'rw', isa => 'Str');

my $context; # keep a weakend copy for the Request class to use

my ($conf_disposition, $conf_uri_class); # configured values

=head1 NAME

Catalyst::Plugin::SmartURI - Configurable URIs for Catalyst

=head1 SYNOPSIS

In your lib/MyApp.pm, load the plugin and your other plugins, for example:

    use Catalyst qw/
        -Debug
        ConfigLoader
        Static::Simple
        Session
        Session::Store::Memcached
        Session::State::Cookie
        Authentication
        Authorization::Roles
        +CatalystX::SimpleLogin
        SmartURI
    /;

In your .conf:

    <Plugin::SmartURI>
        disposition host-header   # application-wide
        uri_class   URI::SmartURI # by default
    </Plugin::SmartURI>

Per request:

    $c->uri_disposition('absolute');

Methods on URIs:

    <a href="[% c.uri_for('/foo').relative %]" ...

=head1 DESCRIPTION

Configure whether C<< $c->uri_for >> and C<< $c->req->uri_with >> return absolute, hostless or
relative URIs, or URIs based on the 'Host' header. Also allows configuring which
URI class to use. Works on application-wide or per-request basis.

This is useful in situations where you're for example, redirecting to a lighttpd
from a firewall rule, instead of a real proxy, and you want your links and
redirects to still work correctly.

To use your own URI class, just subclass L<URI::SmartURI> and set
C<uri_class>, or write a class that follows the same interface.

This plugin installs a custom C<< $c->request_class >>, however it does so in a way
that won't break if you've already set C<< $c->request_class >> yourself, ie. by
using L<Catalyst::Action::REST> (thanks mst!).

There is a minor performance penalty in perls older than 5.10, due to
L<Class::C3>, but only at initialization time.

=head1 METHODS

=head2 $c->uri_for

=head2 $c->req->uri_with

Returns a C<< $c->uri_class >> object (L<URI::SmartURI> by default) in the configured
C<< $c->uri_disposition >>.

=head2 $c->req->uri

Returns a C<< $c->uri_class >> object. If the context hasn't been prepared yet, uses
the configured value for C<uri_class>.

C<< $c->req->uri->relative >> will be relative to C<< $c->req->base >>.

=head2 $c->req->referer

Returns a C<< $c->uri_class >> object for the referer (or configured C<uri_class> if
there's no context) with reference set to C<< $c->req->uri >> if it comes from
C<< $c->req->base >>.

In other words, if referer is your app, you can do
C<< $c->req->referer->relative >> and it will do the right thing.

=head1 CONFIGURATION

In myapp.conf:

    <Plugin::SmartURI>
        disposition absolute
        uri_class   URI::SmartURI
    </Plugin::SmartURI>

=over

=item disposition

One of 'absolute', 'hostless', 'relative' or 'host-header'.  Defaults to
'absolute'.

The special disposition 'host-header' uses the value of your 'Host:' header.

=item uri_class

The class to use for URIs, defaults to L<URI::SmartURI>.

=back

=head1 PER REQUEST

    package MyAPP::Controller::RSSFeed;

    ...

    sub begin : Private {
        my ($self, $c) = @_;

        $c->uri_class('Your::URI::Class::For::Request');
        $c->uri_disposition('absolute');
    }

=over

=item $c->uri_disposition('absolute'|'hostless'|'relative'|'host-header')

Set URI disposition to use for the duration of the request.

=item $c->uri_class($class)

Set the URI class to use for C<< $c->uri_for >> and C<< $c->req->uri_with >> for the
duration of the request.

=back

=head1 EXTENDING

C<< $c->prepare_uri >> actually creates the URI, which you can override.

=cut

sub uri_for {
    my $c = shift;

    $c->prepare_uri($c->next::method(@_))
}

{
    package Catalyst::Request::SmartURI;
BEGIN {
  $Catalyst::Request::SmartURI::AUTHORITY = 'cpan:RKITOVER';
}
{
  $Catalyst::Request::SmartURI::VERSION = '0.039';
}
    use Moose;
    extends 'Catalyst::Request';
    use namespace::clean -except => 'meta';

    sub uri_with {
        my $req = shift;

        $context->prepare_uri($req->next::method(@_))
    }

    sub uri {
        my $req = shift;

        my $uri_class = $context ? $context->uri_class : $conf_uri_class;

        $req->next::method(
            $uri_class->new(
                $req->next::method(@_),
                ($req->{base} ? { reference => $req->base } : ())
            )
        )
    }

    sub referer {
        my $req = shift;

        my $uri_class = $context ? $context->uri_class : $conf_uri_class;

        my $referer   = $req->next::method(@_);

        return $referer if not defined $referer;

        my $base      = $req->base;
        my $uri       = $req->uri;

        if ($referer =~ /^$base/) {
            return $uri_class->new($referer, { reference => $uri })
        } else {
            return $uri_class->new($referer);
        }
    }

    __PACKAGE__->meta->make_immutable;
}

sub setup {
    my $app    = shift;
    my $config =$app->config->{'Plugin::SmartURI'} || $app->config->{smarturi};

    ($conf_uri_class, $conf_disposition) = @$config{qw/uri_class disposition/};
    $conf_uri_class   ||= 'URI::SmartURI';
    $conf_disposition ||= 'absolute';

    eval { Class::Load::load_class($conf_uri_class) };
    Catalyst::Exception->throw(
        message => "Could not load configured uri_class $conf_uri_class: $@"
    ) if $@;

    my $request_class = $app->request_class;

    unless ($request_class->isa('Catalyst::Request::SmartURI')) {
        my $new_request_class = $app.'::Request::SmartURI';

        my $inject_rest = (not $request_class->isa('Catalyst::Request::REST'))
            && eval { Class::Load::load_class('Catalyst::Request::REST') };

        Class::C3::Componentised->inject_base(
            $new_request_class,
            'Catalyst::Request::SmartURI',
            ($inject_rest ?
                'Catalyst::Request::REST' : ()),
            $request_class,
        );
        Class::C3::reinitialize();

        $app->request_class($new_request_class);
    }

    $app->next::method(@_)
}

sub prepare_uri {
    my ($c, $uri)   = @_;
    my $disposition = $c->uri_disposition || $conf_disposition;
    my $uri_class   = $c->uri_class       || $conf_uri_class;
# Need the || for $c->welcome_message, otherwise initialization works fine.

    eval { Class::Load::load_class($uri_class) };
    Catalyst::Exception->throw(
        message => "Could not load configured uri_class $uri_class: $@"
    ) if $@;

    my $res;
    if ($disposition eq 'host-header') {
      $res = $uri_class->new($uri, { reference => $c->req->uri })->absolute;
      my $host = $c->req->header('Host');
      my $port = $host =~ s/:(\d+)$// ? $1 : '';

      if ($port) {
          $port = '' if $c->req->uri->scheme eq 'http'  && $port == 80;
          $port = '' if $c->req->uri->scheme eq 'https' && $port == 443;
      }

      $res->host($host);
      $res->port($port) if $port;
    } else {
      $res = $uri_class->new($uri, { reference => $c->req->uri })->$disposition
    }

    $res
}

# Reset accessors to configured values at beginning of request.
sub prepare {
    my $app = shift;

# Also save a copy of the context for the Request class to use.
    my $c = $context = $app->next::method(@_);
    weaken $context;

    $c->uri_class($conf_uri_class);
    $c->uri_disposition($conf_disposition);

    $c
}

__PACKAGE__->meta->make_immutable;

=head1 SEE ALSO

L<URI::SmartURI>, L<Catalyst>, L<URI>

=head1 AUTHOR

Rafael Kitover, C<< <rkitover at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-catalyst-plugin-smarturi at rt.cpan.org>, or through the web
interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-Plugin-SmartURI>.  I
will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Catalyst::Plugin::SmartURI

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-Plugin-SmartURI>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Catalyst-Plugin-SmartURI>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Catalyst-Plugin-SmartURI>

=item * Search CPAN

L<http://search.cpan.org/dist/Catalyst-Plugin-SmartURI>

=back

=head1 ACKNOWLEDGEMENTS

from #catalyst:

vipul came up with the idea

mst came up with the design and implementation details for the current version

kd reviewed my code and offered suggestions

=head1 TODO

I'd like to extend on L<Catalyst::Plugin::RequireSSL>, and make a plugin that
rewrites URIs for actions with an SSL attribute.

=head1 COPYRIGHT & LICENSE

Copyright (c) 2008 Rafael Kitover

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut

__PACKAGE__; # End of Catalyst::Plugin::SmartURI

# vim: expandtab shiftwidth=4 tw=80:
